msal/src/main/java/com/microsoft/identity/client/tenantprofile/AccountAdapter.java

```diff
@@ -0,0 +1,261 @@
+
+            return result;
+        }
+    }
+
+    private static final CacheRecordFilter orphanAccountFilter = new CacheRecordFilter() {
+
+        private boolean hasNoCorrespondingHomeAccount(@NonNull final ICacheRecord guestRecord,
+                                                      @NonNull final List<ICacheRecord> homeRecords) {
+            // Init our sought value
+            final String guestAccountHomeAccountId = guestRecord.getAccount().getHomeAccountId();
+
+            // Create a List of home_account_ids from the homeRecords...
+            final List<String> homeAccountIds = new ArrayList<String>() {{
+                for (final ICacheRecord cacheRecord : homeRecords) {
+                    add(cacheRecord.getAccount().getHomeAccountId());
+                }
+            }};
+
+            return homeAccountIds.contains(guestAccountHomeAccountId);
+        }
+
+        @Override
+        public List<ICacheRecord> filter(@NonNull final List<ICacheRecord> records) {
+            final List<ICacheRecord> result = new ArrayList<>();
+
+            // An account is orphaned if:
+            // 1. There is no home account which exists which has the same home_account_id
+
+            // First, get the home accounts....
+            final List<ICacheRecord> homeRecords =
+                    filterCacheRecords(
+                            records,
+                            new HomeAccountFilter(false)
+                    );
+
+            // Then, get the guest accounts...
+            final List<ICacheRecord> guestRecords =
+                    filterCacheRecords(
+                            records,
+                            new HomeAccountFilter(true)
+                    );
+
+            // Iterate over the guest accounts and find those which have no associated home account
+            for (final ICacheRecord guestRecord : guestRecords) {
+                if (hasNoCorrespondingHomeAccount(guestRecord, homeRecords)) {
+                    result.add(guestRecord);
+                }
+            }
+
+            return result;
+        }
+    };
+
+    /**
+     * For a supplied List of ICacheRecords, create each root IAccount based on the home
+     * account and then add child-nodes based on any authorized tenants.
+     *
+     * @param allCacheRecords
+     * @return
+     */
+    @NonNull
+    public static List<IAccount> adapt(@NonNull final List<ICacheRecord> allCacheRecords) {
+        // First, get all of the ICacheRecords for home accounts...
+        final List<ICacheRecord> homeCacheRecords = filterCacheRecords(
+                allCacheRecords,
+                new HomeAccountFilter(false)
+        );
+
+        // Then, get all of the guest accounts...
+        final List<ICacheRecord> guestCacheRecords = filterCacheRecords(
+                allCacheRecords,
+                new HomeAccountFilter(true)
+        );
+
+        final List<ICacheRecord> orphanCacheRecords = filterCacheRecords(
```

Comment:
Probably worth explaining what you mean by orphanCacheRecords.... I believe you are using this term to convey a cache record for which their is not corresponding root cache record.  We should probably adopt a consistent terminology Parent/Child.  Local/Foreign.  Root/Projection.  Local/Guest to be able to differentiate between these

Reply:
I've renamed this to `guestCacheRecordsWithNoHomeAccount` -- hopefully this is clearer.

Resolved in af766f7ba4aaf018ef97a2fe592dc50703a65342